pkg/store/storepb/custom.go

```diff
@@ -456,3 +461,119 @@ func CompareLabels(a, b []Label) int {
 func LabelsToPromLabelsUnsafe(lset []Label) labels.Labels {
 	return labelpb.ZLabelsToPromLabels(lset)
 }
+
+// Type alias because protoc-go-inject-field does not support
+// managing imports.
+type syncPool = sync.Pool
+
+// The following were copied/pasted from gogoprotobuf generated code with changes
+// to make it work with sync.Pool / []byte slice.
+func (m *SeriesResponse) Marshal() (dAtA []byte, err error) {
```

Comment:
I know it's a copy-pasta from generated code but this makes my eyes bleed 
```suggestion
func (m *SeriesResponse) Marshal() (data []byte, err error) {
```

Reply:
Thanks for this, I like the attempt of hiding complexity but there are a couple of worrying parts in terms of readability and maintainability of this code:
* `**[]byte` is something we can definitely avoid. I guess we do that to have extra info if this was pooled or not but this goes away if we fix next problem:
* We have mixed responsibility on who is suppose to get bytes from pool. 1 -> user of `SeriesResponse` has to get before and put after send. 2 ->  SeriesResponse itself has to use pool to resize the object.

I can see the reason behind it. But wonder if we can simplify this, by:
1. Making sure pool is only used by SeriesResponse. We "get" buf on Marshal and resize if needed
2. Add "Close" method to series response which puts off the buf when we say "it is closed"?

This way the whole thing is encapsulated: Pool is used only here and we never pass any outside bytes and have this weird double pointer

Reply:
I made it the way it is right now is because ideally the same buffer would be shared between marshaling different responses:

* Get a buffer from sync.Pool
* Pass it to Marshal() via SeriesResponse which can make it larger if necessary
* The original caller i.e. Series() function puts it back - it doesn't know if the underlying slice has been changed

This seems completely natural to me because multiple SeriesResponse structs are for different responses, only the buffer & pool is shared for the actual serialization.

If I understood you correctly, you are proposing moving all of the logic into SeriesResponse itself. Then, we would have something like:
```go
var resp storepb.SeriesResponse
..
resp = storepb.NewSeriesResponse(newSeries, pool)
```

And then in a loop, we would have to do:
```go
resp.Result = newSeries
```

As a result, a connection between the SeriesResponse members and a higher-level function, Series(), would appear. We could add a function to "hide" it but it would be very shallow. I'm not sure if this 2nd option is really better :thinking: 

Given that we will remove this code in a few months hopefully it's probably not worth spending so much energy on minute details such as this :smile: 

If you really think this should be the way then I can make this change but because you've approved this I assume that such a minor possible change isn't a blocker to merge this PR